### ISSUE TITLE
bloom filter memory can explode due to bad estimations

### ISSUE THREAD

**Issue by avikivity** [scylladb/scylla#3720]
Bloom filter memory is allocated to fit the number of partitions. However, except for memtable flushes, the number of partitions is not known in advance
 - compactions can lose partitions due to overwrites or expirations
 - rebuild/bootstrap has an internal combining_reader step with the same effects as compaction
 - repair cannot know in advance how many partitions will be repaired

as a result, we'll generally over-estimate the number of partitions, and as a result over-allocate memory. In extreme cases, we can run out of memory.

**Comment by avikivity** [comment]
Bloom filters can be halved in size if the number of bits is a power of two. To do that, logical-OR the upper half into the lower half and discard the upper half. We'll need to coerce the bloom filter size into a power of two, but that also gets rid of a nasty modulo operation, so it's another win.

Solution outline: re-define the partition count as an upper bound, and adjust it dynamically
 - before sealing the sstable, we know the partition count, so we can resize the bloom filter
 - add a bloom filter memory manager that knows about all bloom filters in the shard, and have it reduce bloom filters when the total size exceeds some fraction of the shard memory (say 5%)

**Comment by raphaelsc** [comment]
On Thu, Aug 23, 2018 at 4:25 AM, Avi Kivity <notifications@github.com>
wrote:

> Bloom filter memory is allocated to fit the number of partitions. However,
> except for memtable flushes, the number of partitions is not known in
> advance
>
>    - compactions can lose partitions due to overwrites or expirations
>
>
yes. that's specially affecting cleanup and tombstone compaction.

**Comment by nyh** [comment]
This is a duplicate or overlaps with several other issues: I found #2024, #2130, #1946, #2465.

Some notes:

@avikivity mentioned "overwrites" as one of the problems, but this specificially is not a problem, because we have a "cardinality estimator" for exactly this purpose. Overwriting by a tombstone and expirations are more of a problem. I'm pretty sure it's possible to come up with a theoretical solution for this problem (like a cardinality estimator which takes into account also deletions), but it's probably not worth the effort, and I like @avikivity 's two ideas, which is important to note are two separate ideas:

1. We can rebuild the the bloom filter after writing an sstable if it's objectively too big (based on what turns out to be a wrong estimate on number of partitions). This rebuild only requires reading the index file, which theoretically is small percentage of the data we just wrote so should not slow down the sstable writing considerably.

2. Also, if the on-disk bloom filter is objectively in the correct size (estimated false-positive rate is our desired rate) but too big for us to hold in memory at the moment (see #2024, #1946 and also the discussion here for various possible reasons why this can happen) we can resample the bloom filter held in memory to reduce memory usage traded-off with increased false positive rate. It's indeed easier to do this if the size is a power of two.

I think the many issues we have on this topic should probably be consolidated into fewer issues which make more specific suggestions (like the two separate ideas above).
